docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,496 @@
+
+1. Explicit client configuration
+2. JVM system properties
+3. Environment variables
+4. AWS profile
+
+*NOTE*: The client needs to be instantiated using `fromEnvironment()` for it to check all the available locations or else it
+will only look at the explicit client config.
+
+Will look in all 4 locations:
+```kotlin
+    val client = S3Client.fromEnvironment {
+        region = "us-west-2"
+    }
+```
+
+Will only look at explicit client config:
+```kotlin
+    val client = S3Client {
+        region = "us-west-2"
+        requestCompression {
+            ...
+        }
+    }
+```
+
+#### Disable Request Compression 
+* Type: Boolean
+* Default: False
+
+
+| Description                  | Client configuration             | JVM system property                  | Environment variable                     | AWS profile key                      |
+|------------------------------|----------------------------------|--------------------------------------|------------------------------------------|--------------------------------------|
+| Enable/disable compression   | `disableRequestCompression`      | `aws.disableRequestCompression`      | `AWS_DISABLE_REQUEST_COMPRESSION`        | `disable_request_compression`        |
+
+#### Minimum Request Compression Size Bytes 
+* Type: Long
+* Default: 10,240 (10KB)
+* Range: 0 - 10,485,760 (0MB - 10MB) 
+* Notes: Setting will be ignored if payload is a stream
+
+| Description                  | Client configuration             | JVM system property                  | Environment variable                     | AWS profile key                      |
+|------------------------------|----------------------------------|--------------------------------------|------------------------------------------|--------------------------------------|
+| Min payload size to compress | `requestMinCompressionSizeBytes` | `aws.requestMinCompressionSizeBytes` | `AWS_REQUEST_MIN_COMPRESSION_SIZE_BYTES` | `request_min_compression_size_bytes` |
+
+This feature can also be configured to use user supplied compression algorithms. The user can add compression algorithms using the client builder. A user supplied algorithm may override the built-in implementation, and the compression algorithm must implement the `CompressionAlgorithm` interface.
+```kotlin
+val client = S3Client.fromEnvironment {
+    region = "us-west-2"
+    requestCompression {
+        compressionAlgorithms = mutableListOf(CustomGzip())
+    }
+}
+
+class CustomGzip : CompressionAlgorithm {
+    override val id: String = "gzip"
+    ...
+}
+```
+
+### The Trait
+Below is an abbreviated example of the request compression trait in use:
+
+```smithy
+@requestCompression(
+    encodings: ["gzip", "custom"]
+)
+operation ...
+```
+
+The client has its own list of supported compression algorithms as a list of `CompressionAlgorithm`s in the client config.
+```kotlin
+val compressionAlgorithms: List<CompressionAlgorithm> = listOf(
+    Gzip()
+)
+```
```

Comment:
Modeling / trait usage is not user-facing and doesn't need to be called out in the design doc imo